internal/pkg/runtime/engines/singularity/process_linux.go

```diff
@@ -146,8 +147,106 @@ func (engine *EngineOperations) checkExec() error {
+	}
+	// the master process does not need this file descriptor after
+	// running the program, make sure it gets closed; ignore any
+	// errors that happen here
+	defer fh.Close()
+
+	// The assumption is that the plugin prepared "Program" in such
+	// a way that it's missing the last parameter and that must
+	// correspond to /dev/fd/N. Instead of making assumptions as to
+	// how many and which file descriptors are open at this point,
+	// simply assume that it's possible to map the existing file
+	// descriptor to the name number in the new process.
+	//
+	// "newFd" should be the same as "fd", but do not assume that
+	// either.
+	newFd := fh.Fd()
+	fdDevice := fmt.Sprintf("/dev/fd/%d", newFd)
+	args := append(program, fdDevice)
+	sylog.Debugf("args=%v", args)
+	cmd := exec.Command(args[0], args[1:]...)
+
+	// Add the /dev/fuse file descriptor to the list of file
+	// descriptors to be passed to the new process.
+	//
+	// ExtraFiles is an array of *os.File, with the position of each
+	// entry determining the resulting file descriptor number. Since
+	// we are passing /dev/fd/N above, place our file handle at
+	// position N-3, so that it gets mapped to file descriptor N in
+	// the new process (the Go library will set things up so that
+	// stdin, stdout and stderr are 0, 1, and 2, so the first
+	// element of ExtraFiles gets 3).
+	cmd.ExtraFiles = make([]*os.File, newFd-3+1)
+	cmd.ExtraFiles[newFd-3] = fh
+	// The FUSE driver will get SIGQUIT if the parent dies.
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		Pdeathsig: syscall.SIGQUIT,
+	}
+
+	if err := cmd.Start(); err != nil {
+		sylog.Debugf("cannot start program %v: %v\n", args, err)
+		return err
+	}
+
+	return nil
+}
+
+// setupFuseDrivers runs the operations required by FUSE drivers before
+// the user process starts
+func setupFuseDrivers(engine *EngineOperations) error {
+	// close file descriptors open for FUSE mount
+	for _, name := range engine.EngineConfig.GetPluginFuseMounts() {
+		var cfg struct {
+			Fuse struct {
+				DevFuseFd int
+				Program   []string
+			}
+		}
+		if err := engine.EngineConfig.GetPluginConfig(name, &cfg); err != nil {
+			return err
+		}
+
+		if err := runFuseDriver(name, cfg.Fuse.Program, cfg.Fuse.DevFuseFd); err != nil {
+			return err
+		}
+
+		syscall.Close(cfg.Fuse.DevFuseFd)
+	}
+
+	return nil
+}
+
+// preStartProcess does the final set up before starting the user's
+// process.
+func preStartProcess(engine *EngineOperations) error {
+	// TODO(mem): most of the StartProcess method should be here, as
+	// it's doing preparation for actually starting the user
+	// process.
+	//
+	// For now it's limited to doing the final set up for FUSE
+	// drivers
+	if err := setupFuseDrivers(engine); err != nil {
+		return err
+	}
+
+	return nil
+}
+
 // StartProcess starts the process
 func (engine *EngineOperations) StartProcess(masterConn net.Conn) error {
+	if err := preStartProcess(engine); err != nil {
```

Comment:
@DrDaveD mentioned an issue with signals, you will need to queue them before executing the FUSE program

Reply:
@cclerget I don't recall exactly what you're referring to, unless you mean the issue of leaving behind running FUSE processes if the singularity container is killed.  I deal with that issue by always enabling --pid.  In fact the [pull request on this branch in mem's repository](https://github.com/mem/singularity/pull/2) that I submitted yesterday automatically enables --pid whenever --fusecmd is used.  With that, I don't see a need for doing anything special with signals here.  The fuse command should put itself into a separate session.

On the other hand, there is still a problem with starting sshfs if it requires prompting for a passphrase.  I think the thing to do there is to simply wait for the process.  Ordinarily a FUSE process should put itself into the background.

Reply:
IIRC there was an issue with defunc processes not correctly catched by PID 1 process ?

Reply:
Good recalling.  I had forgotten about that but quickly discovered it again when I tried using cmd.Run() instead of cmd.Start() for the fuse command.  I think the solution to that is to just have sinit wait for any orphan children before it starts, as I put into [this pull request in mem's repository](https://github.com/mem/singularity/pull/3).

Reply:
I don't know what you guys are talking about...

```
builddir/singularity shell --fusemnt /mnt ./images/fuse-example.sif
WARNING: Cannot find the fuse-example driver in $PATH.
Singularity fuse-example.sif:~/devel/sylabs/singularity/src/github.com/sylabs/singularity> cat /mnt/hello
Hello World!
```

the driver is fuse-example:

```
$ pgrep fuse-example
10377
$ ps `pgrep fuse-example`
  PID TTY      STAT   TIME COMMAND
10377 pts/5    Sl     0:00 /bin/fuse-example -f /dev/fd/5
```

after exiting the shell by typing `exit` and the command line:

```
$ pgrep fuse-example
$
```

the program is properly stopped.

The reason why this is started with `cmd.Start()` and not `cmd.Run()` is because the driver should *not* background itself. If there's a reason why a driver should background itself, we can talk about it after this merges.

Reply:
The fuse drivers that I am familiar with (sshfs & cvmfs) do background themselves so I assume it is normal.  If singularity doesn't wait, then sshfs doesn't work if needs to prompt for a password, as it does when I test it.  Is fuse-example a realistic example?  Where does it come from?

Reply:
By default FUSE program run in background, if foreground is the expected way to run FUSE program, it may be safer to document that somewhere further or better to enforce this mode in the code. Additionally it will be the plugin responsibility to also reset the `--no-init` option in order to always reach the wait process code in `StartProcess` otherwise the condition `(!isInstance && !shimProcess)` will be met when `--no-init` is set, fixable by setting it to `false` in `setupFuseDrivers` loop. I approve this PR as is, feel free to consider or not those suggestions for further work.